fast_node.go

```diff
@@ -24,22 +25,46 @@ func NewFastNode(key []byte, value []byte, version int64) *FastNode {
 }
 
 // DeserializeFastNode constructs an *FastNode from an encoded byte slice.
-func DeserializeFastNode(buf []byte) (*FastNode, error) {
-	val, n, cause := decodeBytes(buf)
+func DeserializeFastNode(key []byte, buf []byte) (*FastNode, error) {
+	ver, n, cause := decodeVarint(buf)
 	if cause != nil {
-		return nil, errors.Wrap(cause, "decoding fastnode.value")
+		return nil, errors.Wrap(cause, "decoding fastnode.version")
 	}
 	buf = buf[n:]
 
-	ver, _, cause := decodeVarint(buf)
+	val, _, cause := decodeBytes(buf)
 	if cause != nil {
-		return nil, errors.Wrap(cause, "decoding fastnode.version")
+		return nil, errors.Wrap(cause, "decoding node.value")
 	}
 
 	fastNode := &FastNode{
+		key: key,
 		versionLastUpdatedAt: ver,
 		value:                val,
 	}
 
 	return fastNode, nil
 }
+
+func (node *FastNode) encodedSize() int {
```

Comment:
The same method of regular node [here](https://github.com/osmosis-labs/iavl/blob/roman/fast-node-get-set/node.go#L358) looks like the following:
```
func (node *Node) encodedSize() int {
	n := 1 +
		encodeVarintSize(node.size) +
		encodeVarintSize(node.version) +
		encodeBytesSize(node.key)
	if node.isLeaf() {
		n += encodeBytesSize(node.value)
	} else {
		n += encodeBytesSize(node.leftHash) +
			encodeBytesSize(node.rightHash)
	}
	return n
}
```

I don't understand what is that extra byte in the beginning for. At first, I though that it might be used for prefix but it looks like `encodedSize` is only used for measuring the value. I don't think we need an extra byte for the null terminator since we encode each field with its own terminator. Could someone explain what is that extra byte meant for, please?

Reply:
TBH I have no idea either. If thing are serializing + deserializing correctly without it, then it should be fine I hope?

(Its very possible this was just always an error, super under-documented code base with up until recently a bus factor of one)